### DISCUSSION TITLE
How to start training again from last checkpoint

### DISCUSSION THREAD

**Discussion by FatimaArshad-DS** [NVIDIA-NeMo/NeMo#4488] [Q&A]
Hi,

My training got interrupted and Im trying to restart training from last checkpoint. However, training starts from the beginning. How do I make it start from last checkpoint?

**Comment by titu1994** [comment]
If you're using exp manager, it has two flags for resuming training v

Without it, you can try loading your checkpoint using load_from_checkpoinr() and then calling trainer.train() again but we haven't tested that

**Reply by yunbin** [reply]
What is involved to restart training using speech_to_text_finetune.py for model type EncDecHybridRNNTCTCBPEModel?  The script states:  "Currently for simplicity of single script for all model types, we only support `init_from_nemo_model` and `init_from_pretrained_model`", but resume from a ckpt file is essential for any model building tool.

**Comment by FatimaArshad-DS** [comment]
Are you talking about this one? create_checkpoint_callback. It is set to True   @titu1994

**Comment by itzsimpl** [comment]
@FatimaArshad-DS see
https://github.com/NVIDIA/NeMo/blob/8a172df69250da231742893387e9a1dd01b16e3d/nemo/utils/exp_manager.py#L197-L200
and
https://github.com/NVIDIA/NeMo/blob/8a172df69250da231742893387e9a1dd01b16e3d/nemo/utils/exp_manager.py#L204-L206

**Comment by itzsimpl** [comment]
Note that you will need to fix eithr
https://github.com/NVIDIA/NeMo/blob/8a172df69250da231742893387e9a1dd01b16e3d/nemo/utils/exp_manager.py#L188-L189
or set
https://github.com/NVIDIA/NeMo/blob/8a172df69250da231742893387e9a1dd01b16e3d/nemo/utils/exp_manager.py#L194-L195
to point to your specific run, that you wish to resume.

**Comment by ericharper** [comment]
We also some documentation here: https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/core/exp_manager.html

**Comment by XuesongYang** [comment]
You could continue training from the latest checkpoint following the steps:
1. go to your checkpoint directory, and rename the latest `a.ckpt` file to `a-last.ckpt` because https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/exp_manager.py#L420 ;
2. override the default YAML config with `++exp_manager.resume_if_exists=True` and run the same training recipe again.

**Reply by titu1994** [reply]
Btw, if this works can you add it to the Nemo docs section for resuming training ?

**Reply by FatimaArshad-DS** [reply]
It starts with loss = nan :/

**Reply by FatimaArshad-DS** [reply]
I installed torch == 1.11.0 and it works. Weird!

**Reply by XuesongYang** [reply]
> Btw, if this works can you add it to the Nemo docs section for resuming training ?

Sure! This trick works for me. Let's wait for @FatimaArshad-DS to further confirm if it works.

**Reply by FatimaArshad-DS** [reply]
This worked!!!! 
However, what happens if training is interrupted before the epoch ends? What that cause any issue? As far as I have observed it trains on already trained data

**Reply by XuesongYang** [reply]
if training is interrupted within an epoch, then no new checkpoint `(a+1)-last.ckpt` would be saved. So you have to load the previous `a-last.ckpt` again and train.

**Reply by FatimaArshad-DS** [reply]
Training on a-last.ckpt starts training on already seen data samples. How do I avoid it?

**Reply by XuesongYang** [reply]
it did not happen to me. Maybe change another random seed? @titu1994 are you aware of this observation?

**Reply by titu1994** [reply]
If you have not turned on shuffle, or have not used shuffle_n in train_ds, then data loader will not shuffle the data (if using tarred dataset).

**Comment by XuesongYang** [comment]
@FatimaArshad-DS Do you have any other questions about this?

**Comment by DevHector-Bit** [comment]
Its possible save checkpoint from Moe models? i got some issues trying to save one the only way i se now its saving weights but doesnt work for resume training
